src/io/image_reader.cpp

```diff
@@ -0,0 +1,159 @@
+        std::stringstream fmt;
+        fmt << "can't open file: " << path;
+        DeepWorks_Assert(false && fmt.str().c_str());
+    }
+
+    cinfo.err = jpeg_std_error(&err);
+    jpeg_create_decompress(&cinfo);
+    jpeg_stdio_src(&cinfo, infile);
+
+    (void) jpeg_read_header(&cinfo, true);
+    (void) jpeg_start_decompress(&cinfo);
+
+    size_t row_stride = cinfo.output_width * cinfo.output_components;
+    JSAMPARRAY buffer = (*cinfo.mem->alloc_sarray)((j_common_ptr) &cinfo, JPOOL_IMAGE, row_stride, 1);
+
+    int width = static_cast<int>(cinfo.output_width);
+    int height = static_cast<int>(cinfo.output_height);
+    int channels = static_cast<int>(cinfo.output_components);
+
+    deepworks::Tensor out_tensor(deepworks::Shape{channels, height, width});
+
+    deepworks::Tensor::Type *dst_data = out_tensor.data();
+    deepworks::Strides tensor_strides = out_tensor.strides();
+
+    size_t h = 0;
+    while (cinfo.output_scanline < cinfo.output_height) {
+        (void) jpeg_read_scanlines(&cinfo, buffer, 1);
+        const size_t h_offset = tensor_strides[1] * h;
+        for (int w = 0; w < width; ++w) {
+            const size_t w_offset = tensor_strides[2] * w;
+            for (int channel = 0; channel < channels; ++channel) {
+                size_t channel_offset = tensor_strides[0] * channel;
+                dst_data[channel_offset + h_offset + w_offset] = buffer[0][w * channels + channel];
+            }
+        }
+        ++h;
+    }
+
+    (void) jpeg_finish_decompress(&cinfo);
+    jpeg_destroy_decompress(&cinfo);
+    fclose(infile);
+    return out_tensor;
+#else
+    DeepWorks_Assert(false && "Couldn't find LIBJPEG");
+#endif
+    return {};
+}
+
+deepworks::Tensor ReadPngFile(std::string_view path) {
+#ifdef HAVE_PNG
+    FILE *infile = fopen(path.data(), "rb");
+    if (!infile) {
+        std::stringstream fmt;
+        fmt << "can't open file: " << path;
+        DeepWorks_Assert(false && fmt.str().c_str());
+    }
+    char header[8];
+    fread(header, 1, 8, infile);
+    if (png_sig_cmp(reinterpret_cast<png_const_bytep>(header), 0, 8)) {
+        std::stringstream fmt;
+        fmt << "File is not a PNG file: " << path;
+        DeepWorks_Assert(false && fmt.str().c_str());
+    }
+    png_structp png_ptr = png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+
+    png_infop info_ptr = png_create_info_struct(png_ptr);
+    if (setjmp(png_jmpbuf(png_ptr))) {
+        DeepWorks_Assert(false && "Error during init_io");
+    }
+    png_init_io(png_ptr, infile);
+    png_set_sig_bytes(png_ptr, 8);
+
+    png_read_info(png_ptr, info_ptr);
+
+    int width = static_cast<int>(png_get_image_width(png_ptr, info_ptr));
+    int height = static_cast<int>(png_get_image_height(png_ptr, info_ptr));
+    int channels = static_cast<int>(png_get_channels(png_ptr, info_ptr));
+
+    png_read_update_info(png_ptr, info_ptr);
+
+    deepworks::Tensor out_tensor(deepworks::Shape{channels, height, width});
```

Comment:
Why not `HWC`, the image is interleaved , isn't it ?

Reply:
Changed